File: internal/decision/decision.go
package decision

import (
	"fmt"
	"time"

	"github.com/huanghe/HomeState/internal/context"
	"github.com/huanghe/HomeState/internal/guardrails"
	"github.com/huanghe/HomeState/internal/model"
	"github.com/huanghe/HomeState/internal/semantic"
)

// Engine makes action decisions based on context + guardrails.
type Engine struct {
	ctx      *context.Engine
	guard    *guardrails.Engine
	semantic *semantic.Registry
	mode     model.RunMode
}

func New(ctxEngine *context.Engine, guardEngine *guardrails.Engine, sem *semantic.Registry, mode model.RunMode) *Engine {
	return &Engine{
		ctx:      ctxEngine,
		guard:    guardEngine,
		semantic: sem,
		mode:     mode,
	}
}

// SetMode changes the run mode.
func (e *Engine) SetMode(m model.RunMode) {
	e.mode = m
}

// GetMode returns the current run mode.
func (e *Engine) GetMode() model.RunMode {
	return e.mode
}

// ShouldAct evaluates whether an action should be taken on an entity.
func (e *Engine) ShouldAct(action, entityID string) model.Decision {
	ctx := e.ctx.Snapshot()
	dec := model.Decision{
		Action:    action,
		Entity:    entityID,
		Timestamp: time.Now(),
	}

	sem := e.semantic.Get(entityID)
	if sem == nil {
		dec.Allowed = false
		dec.Reason = "no semantic mapping for entity"
		dec.Confidence = 0
		return dec
	}

	// Check guardrails
	allowed, reason := e.guard.IsAllowed(ctx, action)
	dec.GuardrailsPassed = allowed

	if !allowed {
		dec.Allowed = false
		dec.Reason = reason
		dec.Confidence = ctx.Rooms[sem.Room].Confidence
		return dec
	}

	// In observe mode, never allow actions
	if e.mode == model.RunModeObserve {
		dec.Allowed = false
		dec.Reason = "observe mode — context updated but no action taken"
		dec.Confidence = ctx.Rooms[sem.Room].Confidence
		return dec
	}

	// In suggest mode, mark as needs confirmation
	if e.mode == model.RunModeSuggest {
		dec.Allowed = false
		dec.Reason = "suggested — awaiting user confirmation"
		dec.Confidence = ctx.Rooms[sem.Room].Confidence
		return dec
	}

	// Auto mode
	dec.Allowed = true
	dec.Confidence = ctx.Rooms[sem.Room].Confidence
	dec.Reason = e.buildReason(ctx, sem)
	return dec
}

func (e *Engine) buildReason(ctx model.ContextState, sem *model.SemanticMapping) string {
	rs := ctx.Rooms[sem.Room]
	if rs.Occupancy {
		return fmt.Sprintf("%s occupied (confidence: %.2f)", sem.Room, rs.Confidence)
	}
	return fmt.Sprintf("%s unoccupied (confidence: %.2f)", sem.Room, rs.Confidence)
}

File: custom_components/homestate/decision.py
"""Decision Layer — action decisions based on context + guardrails."""

from __future__ import annotations

from .context import ContextEngine
from .guardrails import GuardrailsEngine
from .models import Decision, RunMode
from .semantic import SemanticRegistry


class DecisionEngine:
    """Makes action decisions."""

    def __init__(
        self,
        ctx: ContextEngine,
        guard: GuardrailsEngine,
        sem: SemanticRegistry,
        mode: RunMode = RunMode.OBSERVE,
    ) -> None:
        self._ctx = ctx
        self._guard = guard
        self._sem = sem
        self._mode = mode

    @property
    def mode(self) -> RunMode:
        return self._mode

    @mode.setter
    def mode(self, value: RunMode) -> None:
        self._mode = value

    def should_act(self, action: str, entity_id: str) -> Decision:
        ctx = self._ctx.state
        dec = Decision(action=action, entity=entity_id)

        sem = self._sem.get(entity_id)
        if sem is None:
            dec.allowed = False
            dec.reason = "no semantic mapping for entity"
            return dec

        room_state = ctx.rooms.get(sem.room)
        room_confidence = room_state.confidence if room_state else 0.0

        # Check guardrails
        allowed, reason = self._guard.is_allowed(ctx, action)
        dec.guardrails_passed = allowed

        if not allowed:
            dec.allowed = False
            dec.reason = reason
            dec.confidence = room_confidence
            return dec

        if self._mode == RunMode.OBSERVE:
            dec.allowed = False
            dec.reason = "observe mode — context updated but no action taken"
            dec.confidence = room_confidence
            return dec

        if self._mode == RunMode.SUGGEST:
            dec.allowed = False
            dec.reason = "suggested — awaiting user confirmation"
            dec.confidence = room_confidence
            return dec

        # Auto mode
        dec.allowed = True
        dec.confidence = room_confidence
        dec.reason = self._build_reason(ctx, sem.room, room_state)
        return dec

    @staticmethod
    def _build_reason(ctx, room: str, room_state) -> str:
        if room_state and room_state.occupancy:
            return f"{room} occupied (confidence: {room_state.confidence:.2f})"
        return f"{room} unoccupied (confidence: {room_state.confidence:.2f})"
